core/genesis/genesisBlockCreator.go

```diff
@@ -61,21 +73,219 @@ func CreateShardGenesisBlockFromInitialBalances(
+		return nil, process.ErrNilMetaBlockPool
+	}
+
+	argsHook := hooks.ArgBlockChainHook{
+		Accounts:         args.Accounts,
+		AddrConv:         args.AddrConv,
+		StorageService:   args.Store,
+		BlockChain:       args.Blkc,
+		ShardCoordinator: args.ShardCoordinator,
+		Marshalizer:      args.Marshalizer,
+		Uint64Converter:  args.Uint64ByteSliceConverter,
+	}
+	vmFactory, err := metachain.NewVMContainerFactory(argsHook, args.NodesSetup)
+	if err != nil {
+		return nil, err
+	}
+
+	argsParser, err := smartContract.NewAtArgumentParser()
+	if err != nil {
+		return nil, err
+	}
+
+	vmContainer, err := vmFactory.Create()
+	if err != nil {
+		return nil, err
+	}
+
+	interimProcFactory, err := metachain.NewIntermediateProcessorsContainerFactory(
+		args.ShardCoordinator,
+		args.Marshalizer,
+		args.Hasher,
+		args.AddrConv,
+		args.Store,
+		args.MetaDatapool,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	interimProcContainer, err := interimProcFactory.Create()
+	if err != nil {
+		return nil, err
+	}
+
+	scForwarder, err := interimProcContainer.Get(block.SmartContractResultBlock)
+	if err != nil {
+		return nil, err
+	}
+
+	scProcessor, err := smartContract.NewSmartContractProcessor(
+		vmContainer,
+		argsParser,
+		args.Hasher,
+		args.Marshalizer,
+		args.Accounts,
+		vmFactory.BlockChainHookImpl(),
+		args.AddrConv,
+		args.ShardCoordinator,
+		scForwarder,
+		&metachain.TransactionFeeHandler{},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	txTypeHandler, err := coordinator.NewTxTypeHandler(args.AddrConv, args.ShardCoordinator, args.Accounts)
+	if err != nil {
+		return nil, err
+	}
+
+	txProcessor, err := transaction2.NewMetaTxProcessor(
+		args.Accounts,
+		args.AddrConv,
+		args.ShardCoordinator,
+		scProcessor,
+		txTypeHandler,
+	)
+	if err != nil {
+		return nil, process.ErrNilTxProcessor
+	}
+
+	err = deploySystemSmartContracts(txProcessor, vmFactory.SystemSmartContractContainer(), args.AddrConv)
+	if err != nil {
+		return nil, err
+	}
+
+	err = setStakingData(txProcessor, args.NodesSetup.InitialNodesInfo(), args.NodesSetup.StakedValue)
+	if err != nil {
+		return nil, err
+	}
+
+	rootHash, err := args.Accounts.Commit()
+	if err != nil {
+		return nil, err
+	}
+
 	header := &block.MetaBlock{
 		RootHash:     rootHash,
 		PrevHash:     rootHash,
 		RandSeed:     rootHash,
 		PrevRandSeed: rootHash,
 	}
-	header.SetTimeStamp(genesisTime)
+	header.SetTimeStamp(args.GenesisTime)
 
 	return header, nil
 }
 
+// deploySystemSmartContracts deploys all the system smart contracts to the account state
+func deploySystemSmartContracts(
+	txProcessor process.TransactionProcessor,
+	systemSCs vm.SystemSCContainer,
+	addrConv state.AddressConverter,
+) error {
+	tx := &transaction.Transaction{
+		Nonce:     0,
+		Value:     big.NewInt(0),
+		RcvAddr:   make([]byte, addrConv.AddressLen()),
+		GasPrice:  0,
+		GasLimit:  0,
+		Data:      "deploy@" + hex.EncodeToString(factory.SystemVirtualMachine),
```

Comment:
Is the Data field created correctly? For me it crashes on process\smartContract\process.go on method createVMDeployInput line 350

vmCreateInput.ContractCode, err = hex.DecodeString(string(hexCode))

Reply:
also metaDataPool creation is missing the currBlockTxs

 return &metaDataPool{
		metaBlocks:           metaBlocks,
		miniBlocks:           miniBlocks,
		shardHeaders:         shardHeaders,
		headersNonces:        headersNonces,
		transactions:         transactions,
		unsignedTransactions: unsignedTransactions,
		**currBlockTxs:currBlockTxs,**
	}, nil

Reply:
done.

Reply:
ok